File: _posts/2020-12-06/solve.py
#!/usr/bin/env python3
from functools import reduce

groups = open('./input').read().split('\n\n');


def uniq_chars(s):
    char_sets = map(set, s.split('\n'))
    chars = reduce(lambda a, b: a.union(b), char_sets)
    return len(chars)


def common_chars(s):
    char_sets = map(set, s.split('\n'))
    chars = reduce(lambda a, b: a.intersection(b), char_sets)
    return len(chars)


def main():
    print('Total uniq:  ', sum(map(uniq_chars, groups)))
    print('Total common:', sum(map(common_chars, groups)))

main()

File: _posts/2020-12-15/solve.lua
#!/usr/bin/env lua

io.input(io.open("./input", "r"))
ns = {}
lookup = {}
for n in string.gmatch(io.read(), "%d+") do
    n = tonumber(n)
    ns[#ns+1] = n
    lookup[n] = #ns
end

ns[#ns+1] = 0

for i = #ns, 30000000 do
    prev = ns[i]
    curr = lookup[prev]
    if curr == nil then
        curr = 0
    else
        curr = i - curr
    end
    lookup[prev] = i
    ns[i+1] = curr
end

print("2020th: " .. ns[2020])
print("30000000th: " .. ns[30000000])

